analysis-engine/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisEngine.java

```diff
@@ -0,0 +1,161 @@
+  private static final Runnable NO_OP_TERMINATION = () -> {};
+
+  private final GlobalAnalysisContainer globalAnalysisContainer;
+  private final BlockingQueue<AsyncCommand<?>> commandQueue = new LinkedBlockingQueue<>();
+  private final Thread analysisThread = new Thread(this::executeQueuedCommands, "sonarlint-analysis-engine");
+  private final ClientLogOutput logOutput;
+  private final AtomicReference<Runnable> termination = new AtomicReference<>();
+  private final AtomicReference<AsyncCommand<?>> executingCommand = new AtomicReference<>();
+
+  public AnalysisEngine(AnalysisEngineConfiguration analysisGlobalConfig, PluginInstancesRepository pluginInstancesRepository, @Nullable ClientLogOutput logOutput) {
+    globalAnalysisContainer = new GlobalAnalysisContainer(analysisGlobalConfig, pluginInstancesRepository);
+    this.logOutput = logOutput;
+    start();
+  }
+
+  private void start() {
+    // if the container cannot be started, the thread won't be started
+    globalAnalysisContainer.startComponents();
+    analysisThread.start();
+  }
+
+  private void executeQueuedCommands() {
+    while (termination.get() == null) {
+      SonarLintLogger.setTarget(logOutput);
+      try {
+        executingCommand.set(commandQueue.take());
+        if (termination.get() != null && termination.get() != NO_OP_TERMINATION) {
+          executingCommand.get().cancel();
+          break;
+        }
+        executingCommand.get().execute(getModuleRegistry());
+        executingCommand.set(null);
+      } catch (InterruptedException e) {
+        LOG.error("Analysis engine interrupted", e);
+      }
+    }
+    termination.get().run();
+  }
+
+  public <T> CompletableFuture<T> post(Command<T> command, ProgressMonitor progressMonitor) {
+    if (termination.get() != null) {
+      LOG.error("Analysis engine stopping, ignoring command");
+      return CompletableFuture.completedFuture(null);
+    }
+    if (!analysisThread.isAlive()) {
+      LOG.error("Analysis engine not started, ignoring command");
+      return CompletableFuture.completedFuture(null);
+    }
+
+    var asyncCommand = new AsyncCommand<>(command, progressMonitor);
+    try {
+      commandQueue.put(asyncCommand);
+    } catch (InterruptedException e) {
+      asyncCommand.future.completeExceptionally(e);
+    }
+    return asyncCommand.future;
+  }
+
+  public void finishGracefully() {
+    termination.compareAndSet(null, this::honorPendingCommands);
+  }
+
+  private void honorPendingCommands() {
+    List<AsyncCommand<?>> pendingCommands = new ArrayList<>();
+    commandQueue.drainTo(pendingCommands);
+    pendingCommands.forEach(c -> c.execute(getModuleRegistry()));
+    globalAnalysisContainer.stopComponents(false);
+  }
+
+  public void stop() {
+    if (!analysisThread.isAlive()) {
+      return;
+    }
+    if (!termination.compareAndSet(null, NO_OP_TERMINATION)) {
+      // already terminating
+      return;
+    }
+    analysisThread.interrupt();
```

Comment:
What about moving the interrupt after command.cancel(), in order to give more chance to a "clean" termination?